### PR TITLE
chore: Add support for TypeScript files to configure GraphQL servers (`/src/server` folder)

### PR DESCRIPTION
**Pull request by victorhmp**
## What's the purpose of this pull request?

This PR aims to increase the usage of TypeScript in the codebase, and the main feature added was the transition from using `.js` files in the `/src/server` folder to using `.ts` files. This should enable users to extend the default server configuration and GraphQL Schema using TypeScript files, which are not supported out of the box today.

This PR also allows us to use TypeScript in the following Gatsby specific files:

- `gatsby-config`
- `gatsby-ssr`
- `gatsby-browser`

Some minor changes also implemented:

- Added a missing developer dependency: `@lhci/cli`. This should be present in the `package.json` file since our `lhci` script depends on it.
- Renames the `src/components/ThirdyPartyScripts` folder to `ThirdPartyScripts`.
- Fixed type inconsistencies between types defined in `@faststore/sdk` and how they were being used here.

## How to test it?

Since this is more of a structural change, to make sure everything is working, all PR checks should be working and the deploy preview should be no different from what's currently live.

## Checklist

<em>You may erase this after checking them all ;)</em>

- [] CHANGELOG entry added

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -60,6 +60,7 @@
     "@graphql-codegen/cli": "^2.2.1",
     "@graphql-codegen/typescript": "^2.2.4",
     "@graphql-codegen/typescript-operations": "^2.1.8",
+    "@lhci/cli": "^0.9.0",
```
Comment (tlgimenes) [inline package.json:63]: 👀

**Review by tlgimenes — COMMENTED** [review]
i'm not sure about this one. As the PR itself says, Gatsby does not support .ts config files. I'm afraid we get stuck in this decision in the future with future gatsby changes, like Gatsby v4 etc.

One thing I'm not sure works is the config hmr, like ssr and browser configs. For this, try doing this test:
1. yarn develop
2. create a new component
3. Add this new component to wrapRootElement on gatsby-browser
4. Check that this new element is rendered on the screen without restarting the `yarn develop`

File: cypress/global.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import storeConfig from '../store.config'
+import storeConfig from '../src/store.config'
```
Comment (igorbrasileiro) [inline cypress/global.js:1]: Have you ever tried to use nonrelative import? `src/store.config`

File: cypress/integration/a11y.test.js [hunk 3]
```diff
@@ -4,7 +4,7 @@
  * Cypress tests for a11y (accessibility)
  */
 
-import { cypress } from '../../store.config'
+import { cypress } from '../../src/store.config'
```
Comment (igorbrasileiro) [inline cypress/integration/a11y.test.js:7]: Also here, the same nonrelative import.

File: cypress/integration/analytics.test.js [hunk 4]
```diff
@@ -5,7 +5,7 @@
  */
 
 import { options } from '../global'
-import { cypress } from '../../store.config'
+import { cypress } from '../../src/store.config'
```
Comment (igorbrasileiro) [inline cypress/integration/analytics.test.js:8]: Also here, the same nonrelative import.

File: cypress/integration/cart.test.js [hunk 5]
```diff
@@ -5,7 +5,7 @@
  */
 
 import { options } from '../global'
-import { cypress } from '../../store.config'
+import { cypress } from '../../src/store.config'
```
Comment (igorbrasileiro) [inline cypress/integration/cart.test.js:8]: Also here, the same nonrelative import.

File: cypress/integration/performance.test.js [hunk 6]
```diff
@@ -5,7 +5,7 @@
  */
 
 import { options } from '../global'
-import { cypress } from '../../store.config'
+import { cypress } from '../../src/store.config'
```
Comment (igorbrasileiro) [inline cypress/integration/performance.test.js:8]: Also here, the same nonrelative import.

File: cypress/integration/plp.test.js [hunk 7]
```diff
@@ -5,7 +5,7 @@
  */
 
 import { options } from '../global'
-import { cypress } from '../../store.config'
+import { cypress } from '../../src/store.config'
```
Comment (igorbrasileiro) [inline cypress/integration/plp.test.js:8]: Also here, the same nonrelative import.

File: cypress/integration/seo.test.js [hunk 8]
```diff
@@ -7,7 +7,7 @@
  */
 
 import { options } from '../global'
-import { cypress } from '../../store.config'
+import { cypress } from '../../src/store.config'
```
Comment (igorbrasileiro) [inline cypress/integration/seo.test.js:10]: Also here, the same nonrelative import.

File: gatsby-ssr.tsx [hunk 9]
```diff
@@ -1,17 +1,27 @@
...
-import ThirdyPartyScripts from './src/components/ThirdyPartyScripts'
+import ThirdPartyScripts from './src/components/ThirdPartyScripts'
 import Layout from './src/Layout'
 import AnalyticsHandler from './src/sdk/analytics'
 import { validateCart } from './src/sdk/cart/validate'
 import ErrorBoundary from './src/sdk/error/ErrorBoundary'
 import TestProvider from './src/sdk/tests'
 import { uiActions, uiEffects, uiInitialState } from './src/sdk/ui'
-import storeConfig from './store.config'
+import storeConfig from './src/store.config'
```
Comment (igorbrasileiro) [inline gatsby-ssr.tsx:12]: Also here, the same nonrelative import.

File: gatsby-ssr.tsx [hunk 10]
```diff
@@ -1,17 +1,27 @@
...
 
-export const wrapRootElement = ({ element }) => (
+// Gatsby types the returned elements from `getHeadComponents` as
+// `React.ReactNode`, but this is inaccurate. The attributes defined below
+// are present in those elements.
+interface HeadComponents
+  extends ReturnType<PreRenderHTMLArgs['getHeadComponents']> {
+  type: string
+  key: string
+  props?: Record<string, unknown>
```
Comment (igorbrasileiro) [inline gatsby-ssr.tsx:21]: Record<string, unknow> isn't accurate, can pass prop as a number and symbol.

**Comment by birnayt** [comment]
> i'm not sure about this one. As the PR itself says, Gatsby does not support .ts config files. I'm afraid we get stuck in this decision in the future with future gatsby changes, like Gatsby v4 etc.
> 
> One thing I'm not sure works is the config hmr, like ssr and browser configs. For this, try doing this test:
> 
> 1. yarn develop
> 2. create a new component
> 3. Add this new component to wrapRootElement on gatsby-browser
> 4. Check that this new element is rendered on the screen without restarting the `yarn develop`

@tlgimenes based on what you said, do you suggest waiting for the tests before implementing in the project?

**Comment by victorhmp** [comment]
Closing this in favor of https://github.com/vtex-sites/base.store/pull/373, which uses the native TypeScript support added to Gatsby v4.

**Pull request closed without merging**